### ISSUE TITLE
[GCStress=C] Failure in JIT/Methodical/ELEMENT_TYPE_IU/_il_dbgu_fld/_il_dbgu_fld.sh

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#24173]
Linux/arm64, GCStress=C, JitStress=1.

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_flow/

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_tst/31/consoleFull

```
JIT/Methodical/ELEMENT_TYPE_IU/_il_dbgu_fld/_il_dbgu_fld.sh [FAIL]
04:31:50         
04:31:50         Return code:      1
04:31:50         Raw output file:      /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_tst/bin/tests/Linux.arm64.Checked/JIT/Methodical/Reports/JIT.Methodical/ELEMENT_TYPE_IU/_il_dbgu_fld/_il_dbgu_fld.output.txt
04:31:50         Raw output:
04:31:50         BEGIN EXECUTION
04:31:50         /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun _il_dbgu_fld.exe ''
04:31:50         128
04:31:50         failed
04:31:50         Expected: 100
04:31:50         Actual: 1
04:31:50         END EXECUTION - FAILED
04:31:50         Test Harness Exitcode is : 1
04:31:50         To run the test:
04:31:50         > set CORE_ROOT=/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root
04:31:50         > /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_jitstress1_tst/bin/tests/Linux.arm64.Checked/JIT/Methodical/ELEMENT_TYPE_IU/_il_dbgu_fld/_il_dbgu_fld.sh
04:31:50         Expected: True
04:31:50         Actual:   False
04:31:50         Stack Trace:
04:31:50           /mnt/j/workspace/dotnet_coreclr/master/arm64_cross_checked_ubuntu16.04/bin/tests/Linux.arm64.Checked/TestWrappers/JIT.Methodical/JIT.Methodical.XUnitWrapper.cs(88347,0): at JIT_Methodical._ELEMENT_TYPE_IU__il_dbgu_fld__il_dbgu_fld_._ELEMENT_TYPE_IU__il_dbgu_fld__il_dbgu_fld_sh()
```

**Comment by BruceForstall** [comment]
Also failed on Linux/x64, GCStress=C, HeapVerify=1, JitMinOpts=1

https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_ubuntu_gcstress0xc_minopts_heapverify1_flow/129/

**Comment by BruceForstall** [comment]
Also Windows/x64, GCStress=C, JitStress=1

https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_windows_nt_gcstress0xc_jitstress1/132/

**Comment by AndyAyersMS** [comment]
I recall spending quite a bit of time in in the 2.1 endgame trying to repro this locally, but never could (see #17428).

**Comment by AndyAyersMS** [comment]
I'll take another look at this one.

**Comment by AndyAyersMS** [comment]
Not surprisingly, I can't repro the failures. Will keep trying.

**Comment by AndyAyersMS** [comment]
Looks like I can repro the failure case that prints `128`, on Linux x64. It just requires 0xC gc stress. Investigating.

**Comment by AndyAyersMS** [comment]
I think this is a bad test.

The `Main` method pins an object `Test.AA` which has an array field `m_ai`. It then stores addresses of the array elements of `m_ai` into a second array. Later it runs through this second array and dereferences the addresses. But since the source array is not pinned, it can move, and it will if GC kicks in at just the right point.

Presumably the failure is sporadic because GC might not move the object, or might move it but not reuse the storage that is now freed up.

The same issue exists in the matching `i` version. So will fix them both....

**Comment by AndyAyersMS** [comment]
Actually the `i` version already got fixed, somewhere in the mists of time (pre 2010).

**Comment by AndyAyersMS** [comment]
Fixing this should also fix #24464
